presto-kafka/src/test/java/io/prestosql/plugin/kafka/TestKafkaIntegrationPushDown.java

```diff
@@ -0,0 +1,205 @@
+                topicNameOffset);
+
+        queryResult = queryRunner.executeWithQueryId(getSession(), sql);
+        assertEquals(getQueryInfo(queryRunner, queryResult).getQueryStats().getProcessedInputPositions(), 2);
+    }
+
+    @Test
+    public void testTimestampCreateTimeModePushDown()
+    {
+        Pair<Long, Long> timePair = createTimestampTestMessages(topicNameCreateTime);
+        DistributedQueryRunner queryRunner = (DistributedQueryRunner) getQueryRunner();
+        // ">= from_unixtime(%s)" insure including index 2, "< from_unixtime(%s)"  insure excluding index 4;
+        String sql = String.format("SELECT count(*) FROM default.%s WHERE _timestamp >= from_unixtime(%s) and _timestamp < from_unixtime(%s)",
+                topicNameCreateTime, timePair.first() / 1000, timePair.second() / 1000);
+
+        // timestamp_upper_bound_force_push_down_enabled default as false.
+        ResultWithQueryId<MaterializedResult> queryResult = queryRunner.executeWithQueryId(getSession(), sql);
+        assertEquals(getQueryInfo(queryRunner, queryResult).getQueryStats().getProcessedInputPositions(), 998);
+
+        // timestamp_upper_bound_force_push_down_enabled set as true.
+        Session enableTmUpperBoundPushDownSession = Session.builder(getSession())
+                .setSystemProperty("kafka.timestamp_upper_bound_force_push_down_enabled", "true")
+                .build();
+
+        queryResult = queryRunner.executeWithQueryId(enableTmUpperBoundPushDownSession, sql);
+        assertEquals(getQueryInfo(queryRunner, queryResult).getQueryStats().getProcessedInputPositions(), 4);
+    }
+
+    @Test
+    public void testTimestampLogAppendModePushDown()
+    {
+        Pair<Long, Long> timePair = createTimestampTestMessages(topicNameLogAppend);
+        DistributedQueryRunner queryRunner = (DistributedQueryRunner) getQueryRunner();
+        // ">= from_unixtime(%s)" insure including index 2, "< from_unixtime(%s)"  insure excluding index 4;
+        String sql = String.format("SELECT count(*) FROM default.%s WHERE _timestamp >= from_unixtime(%s) and _timestamp < from_unixtime(%s)",
+                topicNameLogAppend, timePair.first() / 1000, timePair.second() / 1000);
+        ResultWithQueryId<MaterializedResult> queryResult = queryRunner.executeWithQueryId(getSession(), sql);
+        assertEquals(getQueryInfo(queryRunner, queryResult).getQueryStats().getProcessedInputPositions(), 4);
+    }
+
+    private QueryInfo getQueryInfo(DistributedQueryRunner queryRunner, ResultWithQueryId<MaterializedResult> queryResult)
+    {
+        return queryRunner.getCoordinator().getQueryManager().getFullQueryInfo(queryResult.getQueryId());
+    }
+
+    private Pair<Long, Long> createTimestampTestMessages(String topicName)
+    {
+        long startTime = -1;
+        long endTime = -1;
+        try (KafkaProducer<Long, Object> producer = testingKafka.createProducer()) {
```

Comment:
I totally do not understand what is happening. Here.
Also why do we need such long sleeps here. How much time can this loop take.

Reply:
Hi losipiuk, it is due to the case that I'm using from_unixtime to build test sql. The precision is second level. I may think a new way.

Reply:
I will use timestamp format instead of "from_unixtime" which do not need to make sure the second precision.